File: src/com/blackoutburst/network/Connection.java
package com.blackoutburst.network;

import com.blackoutburst.game.EntityManager;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class Connection implements PacketUtils {

	public Socket socket = null;
	public BufferedReader in = null;
	public OutputStream out = null;
	
	public void connect() {
		final String IP = "192.168.8.164";
		final short PORT = 25565;
		try {
			socket = new Socket(IP, PORT);
			in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
			out = socket.getOutputStream();
			EntityManager.init(Integer.parseInt(in.readLine()));
			System.out.println("Connected");
		} catch (Exception e) {
			System.err.println("Connection error :"+e);
		}
	}
	
	public void write(String data) {
		if (out == null) return;
		
		Thread connectionThread = new Thread(new Runnable(){
			public synchronized void run(){
				try {
					out.write(data.getBytes(StandardCharsets.UTF_8));
				} catch (Exception e) {
					System.err.println("Error while writing data :"+e);
				}	
			}
		});
		connectionThread.setDaemon(true);
		connectionThread.setName("Connection output thread");
		connectionThread.start();
	}
	
	public void read() {
		if (in == null) return;
		
		try {
			String str = in.readLine();
			if (str == null) {
				close();
				return;
			}
			
			sort(str.charAt(0), str);
		} catch (Exception e) {
			close();
			System.err.println("Error while reading data :"+e);
		}			
	}
	
	public void close() {
		try {
			socket.close();
			in.close();
			out.close();
		} catch (Exception e) {
			System.err.println("Error while closing the connection :"+e);
		}
	}
	
}

File: client/src/main/kotlin/dev/blackoutburst/game/network/Connection.kt
package dev.blackoutburst.game.network

import dev.blackoutburst.game.network.packets.PacketManager
import dev.blackoutburst.game.network.packets.PacketPlayOut
import dev.blackoutburst.game.utils.io
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket

class Connection {

    private var socket: Socket? = null
    private var input: InputStream? = null
    private var output: OutputStream? = null

    private var manager = PacketManager()

    var id = -1

    fun start() {
        try {
            socket = Socket("localhost", 15000).let {
                input = it.getInputStream()
                output = it.getOutputStream()
                it
            }.also {
                io {
                    while (!it.isClosed) {
                        read()
                    }
                }
            }
        } catch (ignored: Exception) {
            close()
        }
    }

    private fun read() {
        try {
            input?.let {
                val idData = it.readNBytes(1)
                if (idData.isEmpty()) {
                    close()
                    return
                }

                val id = manager.getId(idData)
                val size = manager.getSize(id)
                val data = it.readNBytes(size)

                manager.decode(id, data)
            }
        } catch (ignored: Exception) {
            close()
        }
    }

    fun write(packet: PacketPlayOut) {
        try {
            output?.let { out ->
                packet.buffer?.let {
                    io {
                        out.write(it.array())
                        out.flush()
                    }
                }
            }
        } catch (ignored: Exception) {
            close()
        }
    }

    fun close() {
        try {
            input?.close()
            output?.close()
            socket?.close()
        } catch (ignored: Exception) {}
    }
}